SQL: Escaped wildcard (*) not accepted in LIKE

palesz opened the pull request:
For a query like `SELECT name FROM test WHERE name LIKE '%c*'` ES SQL generates an error. `*` is not a special character in a `LIKE` construct and it's expected to not needing to be escaped, so the previous query should work as is.
In the `LIKE` pattern any `*` character was treated as invalid character and the usage of `%` or `_` was suggested instead. But `*` is a valid, acceptable non-wildcard on the right side of the `LIKE` operator.

Fix: #55108

Commit:
SQL: Escaped wildcard (*) not accepted in LIKE

For a query like `SELECT name FROM test WHERE name LIKE ''%c*'` ES SQL
generates an error. `*` is not a special character in a `LIKE` construct
and it's expected to not needing to be escaped, so the previous query
should work as is.
In the LIKE pattern any `*` character was treated as invalid character
and the usage of `%` or `_` was suggested instead. But `*` is a valid,
acceptable non-wildcard on the right side of the `LIKE` operator.

Fix: #55108

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/LikeEscapingParsingTests.java:
```diff
@@ -29,7 +29,7 @@ private String error(String pattern) {
 
     private LikePattern like(String pattern) {
-        Expression exp = null;
+        Expression exp;
```
Please revert.

palesz replied:
Will do. I was wondering what is your convention.

matriv replied:
Generally speaking, I don't have a strong preference for explicitly initialising as `null` but some times it helps for readability. In this case we have a clear if/else statement that assigns a value, so imho is not so important.

palesz commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -270,12 +270,6 @@ public LikePattern visitPattern(PatternContext ctx) {
         if (pattern == null) {
             throw new ParsingException(source(ctx.value), "Pattern must not be [null]");
         }
-        int pos = pattern.indexOf('*');
-        if (pos >= 0) {
-            throw new ParsingException(source(ctx.value),
-                    "Invalid char [*] found in pattern [{}] at position {}; use [%] or [_] instead",
-                    pattern, pos);
-        }
```
@costin Did I miss anything, is there a reason to keep this check here?

costin replied:
To prevent folks from using `*` instead of `%`. Whether that's still a concern I don't know.

matriv replied:
We don't have anything in the docs regarding the `*` and we list only the `%` and `_` 
`*` should be interpreted as a normal char for SQL:
```
postgres=# create table mytable (str varchar(10));
CREATE TABLE

postgres=# insert into mytable values('*test*'), ('as*test*fd');
INSERT 0 2

postgres=# select * from mytable where str like ('%*test*%');
    str
------------
 *test*
 as*test*fd
(2 rows)

postgres=# select * from mytable where str like ('*test*');
  str
--------
 *test*
(1 row)
```

How about adding a NOTE in the docs to explicitly specify that `*`, `?` and the usual regex chars are interpreted as normal chars?

palesz replied:
AFAIK it is only the MS Access SQL flavour that treats `*` as a wildcard in a `LIKE` pattern, so SQL folks should be fairly used to `*` not being a wildcard. ES customers with less SQL knowledge might not. I think adding a NOTE to the docs does not hurt.

palesz replied:
I am also thinking about removing the restriction on not being able to use `*` as an `ESCAPE` character from (here)[https://github.com/elastic/elasticsearch/blob/50fc9d52097cab6ac8b55625fc8b8f387688de13/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java#L285]. Do you see any good reason to keep that restriction?
As a user `*` would not be my first choice as an `ESCAPE` character, but the parser can handle it since `*` is just a normal character.

palesz replied:
On PostgreSQL, MySQL, SQLite any character can be used as ESCAPE (even wildchars) and any character can be escaped, not just wildcards. PostgreSQL treats the `'%%' ESCAPE '%'` as a non-wildcard `%` character (just as another escaped character), while MySQL and SQLite treat it as a wildcard (special escaping case). I assume it was intentional to behave differently and only allow escaping of the wildchars.

matriv replied:
As I see it, since only `%` and `_` are wildcards for `LIKE` I would only allow escaping of those, but if we go one with this change it shouldn't be part of the bugfix that would be backported, but for next minor release since it's a minor "breaking" change.

palesz replied:
Agree @matriv that allowing escaping any characters should be a separate PR and likely should be in the next minor release only, although technically it is less of a breaking change unless someone depends on the exceptions around escaping non-wildchars. Anyways, this is off topic for this PR.

To clarify: in this current PR I am only removing the special treatment of `*` and the restriction on using `*`as an `ESCAPE` character (wildchars are still not allowed as `ESCAPE` characters). Does this sound good to you?

matriv replied:
Sorry for the confusion, but if `*` is not a wildcard anymore, maybe worths removing it as escape character as well in this fix? @costin what do you think?

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/parser/LikeEscapingParsingTests.java:
```diff
@@ -63,8 +63,15 @@ public void testInvalidChar() {
+    public void testCannotUseStarWithEscaping() {
         assertThat(error("'|*string' ESCAPE '|'"),
-                is("line 1:11: Invalid char [*] found in pattern [|*string] at position 1; use [%] or [_] instead"));
+            is("line 1:11: Pattern [|*string] is invalid as escape char [|] at position 0 can only escape wildcard chars; found [*]"));
```
Should we maybe improve the error message by listing the wildcard chars for LIKE: `%` and `_` ?

palesz replied:
> Looks good, Maybe add one more test with a more complicated string, something like:
> `*%*_string*_**%`. Left one more comment about the error message.

Will add a new test case, but I will add it to the [LikeConversionTests](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/util/LikeConversionTests.java), since I think that has less to do with `ESCAPE`ing.

costin reviewed (commented):
Please do use Markdown as in the initial report (even more so when copy-pasting) - it improves readability a lot.

matriv reviewed (commented):
Looks good, Maybe add one more test with a more complicated string, something like:
`*%*_string*_**%`. Left one more comment about the error message.

palesz commented:
Sorry for the `force-push`, I should not have done that.

Commit:
Merge branch 'master' of github.com:elastic/elasticsearch into fix/55108

matriv commented on docs/reference/sql/functions/like-rlike.asciidoc:
```diff
@@ -33,8 +33,8 @@ with the `LIKE` operator:
 
-The percent sign represents zero, one or multiple characters. The underscore represents a single number or character. These symbols can be
-used in combinations.
+No other characters have special meaning or act as wildcard. The percent sign represents zero, one or multiple characters.
```
Personally I'd add it as a `[NOTE]` below the existing sentence, so that it stands out.

Commit:
Merge remote-tracking branch 'origin/master' into fix/55108

Commit:
Documentation changes for SQL `LIKE` as per PR

Added a NOTE explaining that only `%` and `_` treated as wildcards
and other characters (even `*` or `?`) are just treated as normal
characters in the `LIKE` pattern.

Commit:
Fixed the checkstyle errors

matriv reviewed (approved):
LGTM. Thx!

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -288,7 +282,7 @@ public LikePattern visitPattern(PatternContext ctx) {
                 // these chars already have a meaning
-                if (escape == '*' || escape == '%' || escape == '_') {
+                if (escape == '%' || escape == '_') {
                     throw new ParsingException(source(escapeCtx.escape), "Char [{}] cannot be used for escaping", escape);
```
On the same idea (which I like) of actually[ mentioning the wildcard chars in the error message itself](https://github.com/elastic/elasticsearch/pull/63428/files#diff-3c0e68f5f3ff479d4817a66493f2d7decb2974c2368cb68bf2bae612b2b91743R301) (for the lazy user who doesn't read the docs and asking "why % char cannot be used for escaping?"), what about adapting this error message to include the reason? Something like `Char [%] cannot be used for escaping as it's one of the wildcard chars [%_]`.

astefan reviewed (approved):
Too late with the review. Sorry about that. Fwiw LGTM, with a minor cosmetic comment. Thanks.
